### ISSUE TITLE
target es2015 modules as well as a bundled script

### ISSUE THREAD

**Issue by toshok** [jquery/esprima#1599]
I'd like to consume esprima using es2015 import syntax (and for echojs I _have_ to, since it doesn't understand `require`/`exports.*` at all).

I've been dealing with this by just fixing esprima to a pretty old version and modifying the bundle to use es2015 exports (https://github.com/toshok/esprima/blob/es6/esprima-es6.js), but I'd like to start tracking closer to HEAD on master, and potentially drop my fork altogether.

**Comment by toshok** [comment]
re: this being a decision about what target to actually distribute, I think it's in keeping with others' views on es2015 in libraries - that the bundling/converting to es5 is something that should happen downstream.

So in that regard I'd support moving esprima to _only_ build es2015 modules.

On the other hand, that's a pretty major api break, and I'd rather do the minimal non-breaking change that makes es2015 modules available for use as soon as possible.

**Comment by ariya** [comment]
Is there a precedent that we can follow? Is there another similar library that already follows this pattern?

Also, do we need to worry about e.g. ES2016, ES2017, and other flavors? I don't expect Esprima will need to use exponentiation (ES2016), but who knows about object rest/spread (ES2017) or other useful future syntax?

What shall we do w.r.t testing?

**Comment by mikesherov** [comment]
@ariya on the projects that I contribute to at work, we target node LTS plus evergreen browsers - 1 version for untranspiled JS source, with an es5 dist. 

So at this point, that would say safe to release an es5 dist, and es2015 version of the source unbundled. Once evergreen + node LTS is es2016, drop the es2015 source and leave a copy of the es2016 source.

**Comment by ariya** [comment]
Is there any convention on `package.json` to designate the es2015/es2016 flavor of the library?

How about the module format? Do we keep the CommonJS variant for es2015/es2016 output? Is it worth to wait for the upcoming native module support in core Node.js?

**Comment by mikesherov** [comment]
Some people have been using `module` instead of `main` to indicate a module entry point but probably best to wait on what Node will do officially in this respect. However, that doesn't prevent us from publishing es2015 code, just from specifying it in package.json.

In the meantime, id think transpiling to es2015 would mean having import statements instead of commonjs require.

**Comment by toshok** [comment]
yeah, part of my hope for es2015 is import/export statements (echojs requires them).

There's also the possibility of bundling the es2015 (using `rollup`, eg).  Rollup can bundle es2015 modules into a single module, also generate es5 (without the many-functions approach to dealing with `require` that webpack uses).  The size is pretty competitive:

```
-rw-r--r--   1 toshok  staff  265029 Oct 23 20:04 esprima.js
-rw-r--r--   1 toshok  staff  235254 Oct 27 10:05 rolledup-cjs.js
-rw-r--r--   1 toshok  staff  235072 Oct 27 10:05 rolledup-es2015.js
```

Definitely not a requirement, though, and the argument could easily be made that bundling should be pushed as far downstream as possible to remain flexible.  It would also represent a user-visible to the existing es5 bundle.

**Comment by ariya** [comment]
I think bundling es5 and es2015 in the `dist` will work for now, at least as a transition step.

What we need to figure out is how to tweak the test infrastructure to accomodate the es2015 flavor as we can't just dump the es2015 version there without making it subjected to the same amount of battery tests.

**Comment by mikesherov** [comment]
Would be interesting to see if the es2015 version suddenly fails. Would that just mean a bug in the typescript transpile?

**Comment by ariya** [comment]
I think I have a possible test strategy. Let's continue to keep the browser-based tests to use the ES5 version, since we're bound to still support some versions of IE, Android browsers, etc. Meanwhile, let's switch the console-based tests (with Node.js >= 6) to use the ES.Next version.

**Comment by mikesherov** [comment]
@ariya SGTM.

**Comment by toshok** [comment]
@ariya likewise, SGTM.  I can try and find some time this week (if not during the week, definitely the weekend) to do that work in a new PR if you'd like.

**Comment by ariya** [comment]
@toshok Sure thing, please do that. Also, I don't think we need two separate `compile` versions, let's have `npm run compile` produces _both_ flavors (ES5 and ES2015).

**Comment by ariya** [comment]
No activity so far, so I'm closing it. We can reopen it again once we revisit the subject.
